Source/Particles/Collision/BinaryCollision.H

```diff
@@ -144,14 +158,50 @@ public:
     void doCollisionsWithinTile (
         int const lev, amrex::MFIter const& mfi,
         WarpXParticleContainer& species_1,
-        WarpXParticleContainer& species_2)
+        WarpXParticleContainer& species_2,
+        amrex::Vector<WarpXParticleContainer*> product_species_vector)
     {
         using namespace ParticleUtils;
         using namespace amrex::literals;
 
         int const ndt = m_ndt;
         FunctorType binary_collision_functor = m_binary_collision_functor;
 
+        // Get product species soa and positions in the form of DeviceVectors
+        const int n_product_species = m_product_species.size();
+        amrex::Gpu::DeviceVector<SoaData_type> soa_product_vector;
+        amrex::Gpu::DeviceVector<GetParticlePosition> get_position_product_vector;
+        constexpr int getpos_offset = 0;
+#ifdef AMREX_USE_GPU
+        amrex::Gpu::HostVector<SoaData_type> host_soa_product_vector;
+        amrex::Gpu::HostVector<GetParticlePosition> host_get_position_product_vector;
+        for (int i = 0; i < n_product_species; i++)
+        {
+            ParticleTileType& ptile_product = product_species_vector[i]->ParticlesAt(lev, mfi);
+            host_soa_product_vector.push_back(ptile_product.getParticleTileData());
+            host_get_position_product_vector.push_back(GetParticlePosition(ptile_product,
+                                                                           getpos_offset));
+        }
+        soa_product_vector.resize(n_product_species);
+        get_position_product_vector.resize(n_product_species);
+        amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice, host_soa_product_vector.begin(),
+                              host_soa_product_vector.end(), soa_product_vector.begin());
+        amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice, host_get_position_product_vector.begin(),
+                              host_get_position_product_vector.end(),
+                              get_position_product_vector.begin());
+        amrex::Gpu::streamSynchronize();
+#else
+        for (int i = 0; i < n_product_species; i++)
+        {
+            ParticleTileType& ptile_product = product_species_vector[i]->ParticlesAt(lev, mfi);
+            soa_product_vector.push_back(ptile_product.getParticleTileData());
+            get_position_product_vector.push_back(GetParticlePosition(ptile_product,
+                                                                      getpos_offset));
+        }
+#endif
+        auto soa_product_data = soa_product_vector.data();
+        auto get_position_product_data = get_position_product_vector.data();
```

Comment:
I think I'm going to need some help for this part of the code.
Basically, I'm trying to pass a runtime dependent number of particle structs of arrays (of type `WarpXParticleContainer::ParticleTileType::ParticleTileDataType`) to GPU kernels (the kernels are started below). I thought I'd use a `DeviceVector` to store the structs of arrays and this is the code I came up with (it feels overly complicated, so maybe there is a much simpler solution altogether). With gcc, this compiles for both CPUs and GPUs (with cuda), although I haven't tried to actually access the soa data from the kernels yet (I want to do this in a follow-up PR).

However, with clang, this code does not compile, at least for CPUs. The error occurs as I try to call the `push_back` method of an `amrex::Gpu::DeviceVector<SoaData_type>` object. I attach here the full error message in case it is useful: [clang_compile_error.txt](https://github.com/ECP-WarpX/WarpX/files/7041992/clang_compile_error.txt). I also cannot assign (neither copy assignment nor move assignment) to an existing object of type `WarpXParticleContainer::ParticleTileType::ParticleTileDataType`, so stuff like `soa_product_vector[i] = ptile_product.getParticleTileData()` does not compile either (with a similar error message). 
Calling `push_back` with an object of type `std::vector<SoaData_type>` seems to compile on the other hand, but I guess that won't work with GPUs.

Does someone know how I could achieve what I want? @atmyers ?

Reply:
I'm not 100% sure why clang doesn't like this code, but as a workaround, could you use `amrex::Vector` instead of `amrex::Gpu::HostVector` for the host container and see if that works?

Reply:
Thanks for the suggestion! Looks like the code compiles with clang with your workaround!